Fix Daylighting Map Output Header and Remove Maximum Reference Points (of 2) Variable

RKStrand opened the pull request:
Pull request overview
---------------------
 - Fixes #8846 
 - A while back, daylighting was updated to allow for more than two reference points.  Unfortunately, the code modifications to remove this limit missed a couple of places where the parameter MaxRefPts was used.  Since this parameter was set to 2, there were still some limits within the daylighting simulation and output reporting.  This has now been fixed and the variable MaxRefPts has been completely removed from the code.  Results show no actual differences except in one file: SolarShadingTestGPU (where the differences are all pretty small and this is a file that seems to have bounced around a bit recently so I don't think these changes are causing the problem).  Note that the differences in the three MAP files are intentional and only in the headers where a separate bug was uncovered while fixing the problem.  These differences are expected.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect
 - [X] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [x] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] ~If feature, test running new feature, try creative ways to break it~
 - [x] CI status: all green or justified
 - [x] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [x] Check any new function arguments for performance impacts
 - [ ] ~Verify IDF naming conventions and styles, memos and notes and defaults~
 - [ ] ~If new idf included, locally check the err file and other outputs~

RKStrand committed:
Corrections to Number of Map and Reference Points for Daylighting

When the number of daylighting map and reference points was expanded, not everything was update and some variables were still using the old hardwired limit of 2.  This code corrects it, gets rid of the old integer parameter completely, cleans up some code, and provides a new unit test.

RKStrand committed:
Merge branch 'develop' into 8846DataDaylightingMaxRefPointsObsolete

RKStrand committed:
Daylighting Ref and Map Points formatting changes

Forgot to run the format check before the commit.

RKStrand committed:
Daylight Map Correction to Compiler Error and E+ Crashes

During the CI runs for the last commit, it was discovered that there were a couple of crashes and a compiler error that came back.  These changes address those issues.

RKStrand committed:
Merge branch 'develop' into 8846DataDaylightingMaxRefPointsObsolete

Myoldmopar commented on src/EnergyPlus/DataDaylighting.hh:
```diff
@@ -65,7 +65,6 @@ namespace EnergyPlus {
 namespace DataDaylighting {
 
     // Two kinds of reference points: used directly in daylighting, used to show illuminance map of zone
-    constexpr int MaxRefPoints(2);       // Maximum number of daylighting reference points, 2
```
:+1:

Myoldmopar commented on src/EnergyPlus/DaylightingManager.cc:
```diff
@@ -1339,6 +1339,8 @@ void CalcDayltgCoeffsMapPoints(EnergyPlusData &state, int const ZoneNum)
 
         if (state.dataDaylightingData->IllumMapCalc(MapNum).Zone != ZoneNum) continue;
 
+        numRefPts = state.dataDaylightingData->IllumMapCalc(MapNum).TotalMapRefPoints;
```
Dynamic assignment of the actual number of reference points.  Good.

Myoldmopar commented on src/EnergyPlus/DaylightingManager.cc:
```diff
@@ -1360,58 +1362,54 @@ void CalcDayltgCoeffsMapPoints(EnergyPlusData &state, int const ZoneNum)
-        // MapWindowSolidAngAtRefPt.allocate( NRF, ZoneDaylight( ZoneNum ).NumOfDayltgExtWins ); //Inactive
-        // MapWindowSolidAngAtRefPtWtd.allocate( NRF, ZoneDaylight( ZoneNum ).NumOfDayltgExtWins ); // Not an array anymore
 
-        for (IL = 1; IL <= NRF; ++IL) {
```
Got it, NRF was already a dynamic variable for the amount of reference points, but now it's just not needed since we already have `numRefPts`.  :+1:

Myoldmopar commented on src/EnergyPlus/DaylightingManager.cc:
```diff
@@ -9955,7 +9953,8 @@ void ReportIllumMap(EnergyPlusData &state, int const MapNum)
         EnvrnPrint.dimension(state.dataDaylightingData->TotIllumMaps, true);
-        RefPts.allocate(state.dataGlobal->NumOfZones, DataDaylighting::MaxRefPoints);
+        RefPts.allocate(state.dataGlobal->NumOfZones,
+                        state.dataDaylightingData->ZoneDaylight(state.dataDaylightingData->IllumMap(MapNum).Zone).TotalDaylRefPoints);
```
Proper allocation :+1:

mjwitte replied:
Um, no, this isn't quite right. This allocation only happens once and will be based on the number of daylighting reference points in the zone of the first map. 

Looking at this more, `RefPts` appears to be used only in this function and only to build a title string for the map output. Why isn't this just a local string variable? And we really don't need to have `RefPts` and `refPt2`, just build the concatenated string directly from the base data in `ZoneDaylight`. I think `RefPts` can be removed from dataDaylightingData and just become a simple local string in this function.

RKStrand replied:
Hmmm...I missed that RefPts was only being used to pass the title field to the report.  So, from what I am seeing, it looks like it can be eliminated from the code and replace with a single character string that can be created and then passed downstream.  I'll try to get that fixed soon (today or tomorrow hopefully).

RKStrand replied:
@mjwitte Just committed a new version that has completely eliminated RefPts.  Hopefully ci comes back clean.

Myoldmopar commented on src/EnergyPlus/DaylightingManager.cc:
```diff
@@ -9999,14 +9998,20 @@ void ReportIllumMap(EnergyPlusData &state, int const MapNum)
     if (EnvrnPrint(MapNum)) {
+        std::string refPt2;
+        if (state.dataDaylightingData->ZoneDaylight(state.dataDaylightingData->IllumMap(MapNum).Zone).TotalDaylRefPoints > 1) {
+            refPt2 = RefPts(state.dataDaylightingData->IllumMap(MapNum).Zone, 2);
```
What's the magic 2 here?  Do we need a more dynamic daylighting map that can handle > 2 now as well?

mjwitte replied:
If I remember correctly, this is just for reference in the map header so you can see where your reference points would fall within the map grid.
```
Date/Time,DAYLIT ZONE:CHICAGO ANN CLG 1% CONDNS DB=>MWB:DAYLIT MAP at 0.80m Illuminance [lux] (Hourly),RefPt1=(2.50:5.00:0.80),RefPt2=(2.50:3.00:0.80),,
 07/21 05:00,(0.10;0.10)=,(0.63;0.10)=,(1.17;0.10)=,(1.70;0.10)=,(2.23;0.10)=,(2.77;0.10)=,(3.30;0.10)=,(3.83;0.10)=,(4.37;0.10)=,(4.90;0.10)=
(0.10;0.10)=,0,0,0,0,0,0,0,0,0,0
(0.10;1.19)=,0,0,0,0,0,0,0,0,0,0
```
So, we could leave it just listing the first two ref points, but it would be better to list all of them.

RKStrand replied:
@Myoldmopar @mjwitte Ok, I've committed another code mod that should address this so that all of the points are listed even if there are more than two.  Also cleaned up the transition rule document omission from a recent PR and something I missed in another documentation only PR.  But, hold on a little bit...I forgot that unit test was still running on my machine and I just got a clang-format issue.  Hoping the unit test comes back clean.  If it does, I'll fix the formatting issue and make another commit.  Hopefully this will be good to go then?

RKStrand replied:
@Myoldmopar @mjwitte Unit test came back clean.  Clang-format has now been run (cleaned up one minor issue) and CI is now running.

Myoldmopar commented on tst/EnergyPlus/unit/DaylightingManager.unit.cc:
```diff
@@ -2887,3 +2888,52 @@ TEST_F(EnergyPlusFixture, DaylightingManager_TDD_NoDaylightingControls)
+    EXPECT_EQ(ExpectedResult1, state->dataDaylightingManager->RefPts(1, 1));
+    EXPECT_EQ(ExpectedResult2, state->dataDaylightingManager->RefPts(1, 2));
+    EXPECT_EQ(ExpectedResult3, state->dataDaylightingManager->RefPts(1, 3));
+}
```
:+1:

Myoldmopar reviewed (commented):
OK, so I _think_ this is good to go, but I do want to know a little about the magic 2 in the daylighting map reporting.  From what I see, it looks like it is still a limitation on the number of possible reference points to report.  Or something.  Which would make a change well beyond scope for this defect, I just want to get my head around it before merging this.  Thanks!

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 8846DataDaylightingMaxRefPointsObsolete

Myoldmopar commented:
I went ahead and pushed an updated branch up, if we can resolve the discussion above this should be ready to go.

Myoldmopar commented:
Windows failed for an unrelated reason, and Mac decided not to respond.  I'd still like to have a defect file to test on per [my comment](https://github.com/NREL/EnergyPlus/issues/8846#issuecomment-882064007) over on the issue.  So I'll probably pull develop in one more time before this merges.

RKStrand commented:
@Myoldmopar I have a file with three reference points that I used for testing.  I've added two files that I was using for testing in a new folder in EnergyPlusDevSupport.  One has three reference points and another has only one.  Hope this helps!

RKStrand committed:
Daylight Ref Pts Title Fix

Updated the title of the daylight map report to show all fields, not just the first two.  Also, to clean up two previous PRs, added language to the transition rules document and fixed something missed on a recent documentation fix. :)

RKStrand committed:
Daylight Manager Change Clang Format Fix

Formatting fix courtesy of clang-format...

mjwitte commented on src/EnergyPlus/DaylightingManager.cc:
```diff
@@ -9999,14 +9998,26 @@ void ReportIllumMap(EnergyPlusData &state, int const MapNum)
                               state.dataEnvrn->EnvironmentName,
                               state.dataDaylightingData->IllumMap(MapNum).Zone,
                               RefPts(state.dataDaylightingData->IllumMap(MapNum).Zone, 1),
-                              RefPts(state.dataDaylightingData->IllumMap(MapNum).Zone, 2),
+                              refPt2,
```
This could be a single parameter. If I'm following correctly, these just get concatenated anyway.

RKStrand committed:
Merge branch 'develop' into 8846DataDaylightingMaxRefPointsObsolete

RKStrand committed:
Removal of RefPts array

Based on comments during the review, the array RefPts was eliminated from the code and replaced with a string variable that is used in headers.  Various other changes were made downstream and in unit tests.  This should hopefully address the concern.

mjwitte commented on src/EnergyPlus/DataDaylighting.hh:
```diff
@@ -234,6 +233,7 @@ namespace DataDaylighting {
         SharedFileHandle mapFile;     // Unit number for map output (later merged to final file)
         bool HeaderXLineLengthNeeded; // X header will likely be the longest line in the file
         int HeaderXLineLength;        // actual length of this X header line
+        std::string pointsHeader;     // part of the header that lists the map points
```
This should say `// part of the header that lists the reference points in the same zone`

mjwitte commented on src/EnergyPlus/DaylightingManager.cc:
```diff
@@ -9984,29 +9980,33 @@ void ReportIllumMap(EnergyPlusData &state, int const MapNum)
+    state.dataDaylightingData->IllumMap(MapNum).pointsHeader = "";
+    for (R = 1; R <= state.dataDaylightingData->ZoneDaylight(state.dataDaylightingData->IllumMap(MapNum).Zone).TotalDaylRefPoints; ++R) {
+        state.dataDaylightingData->IllumMap(MapNum).pointsHeader +=
+            format("RefPt{}=({:.2R}:{:.2R}:{:.2R})",
```
I would add a space before Refpt to make the result a little more readable. `format(" RefPt . . . `
Current output is:
`Date/Time,WEST ZONE:CHICAGO_IL_USA ANNUAL HEATING 99% DESIGN CONDITIONS DB:MAP1 at 0.00m Illuminance [lux] (Hourly),RefPt1=(3.05:3.05:0.90),RefPt2=(3.05:3.05:0.00),RefPt3=(1.05:3.05:0.90),RefPt4=(3.05:1.05:0.00),,`

Also note there are two trailing commas for some reason. I looked at `pointsHeader` in the debugger, and the trailing commas are not here, so they're getting added later somewhere.
Oh, wait, there's a comment from long ago `// must add correct number of commas at end`. Ok, whatever.
Its odd that the z-coordinate of the map gets written to the sql output, but not to the map file. But looking back in the archive, it's always been that way. So, whatever.

So, the only change requested here is to add a space to " RefPt".

mjwitte commented on src/EnergyPlus/SQLiteProcedures.cc:
```diff
@@ -1827,9 +1822,8 @@ void SQLite::createSQLiteDaylightMapTitle(int const mapNum,
-        sqliteBindText(m_daylightMapTitleInsertStmt, 6, refPt2);
-        sqliteBindDouble(m_daylightMapTitleInsertStmt, 7, zCoord);
+        sqliteBindText(m_daylightMapTitleInsertStmt, 5, refPts);
+        sqliteBindDouble(m_daylightMapTitleInsertStmt, 6, zCoord);
```
Realizing now that this changes the number of columns in the SQLite output. @jmarrec Is this ok?
It also puts 1:n daylighting reference point coordinates (which are just informative here, they don't really apply to the map values) into a single string field. e.g.
 "RefPt1=(2.50:2.00:0.80),RefPt2=(2.50:18.00:0.80),RefPt3=(2.50:18.00:0.50)" where before only the first two were written and they were in separate fields.

jmarrec replied:
It'd be great to have the OutputRulesChanges filled out so I could see what's going on a bit better. But you shouldn't be too concerned about changing stuff in the SQL, OpenStudio can adapt to the new format (provided we're aware of the need to do it)

When merging this, it'd be great to open an Enhancement Request on https://github.com/NREL/OpenStudio/issues to clearly document the need to adjust OpenStudio once it targets the next version of EnergyPlus (probably 9.6.0-IOFreeze) though, as we could easily miss it later down the road.

mjwitte reviewed (commented):
@RKStrand Very close.
This will need some notes in the OutputRules doc to explain both the change in the Map.* output and the SQLite output. But let's wait for @jmarrec to confirm if the SQL changes are acceptable.

RKStrand committed:
Corrections to Daylight Map Header

Potentially final changes to address review comments.  Updates to variable comment, spacing in MAP header, Output Rules, and two unit tests.  If this comes back clean, it should be good to go.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 8846DataDaylightingMaxRefPointsObsolete

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 8846DataDaylightingMaxRefPointsObsolete

mjwitte commented:
Fixed conflict in input rules file. Merging.
